### DISCUSSION TITLE
Emit typescript interfaces?

### DISCUSSION THREAD

**Discussion by arabold** [microsoft/typespec#3307] [Q&A] [answered]
I'm able to successfully create a OpenAPI spec from my TypeSpec, as well as JSON Schema files. But how about something as simple as plain TypeScript interfaces? Is there a default emitter that can create these?

I'm aware that I could do TypeSpec -> JSON Schema -> TypeScript, and I tried that already, but that seems a bit convoluted 🤨 

```tsp
// main.tsp
model Pet {}
model Person {
  pet: Pet;
}
```

would become
```ts
// models.d.ts
export interface Pet {}
export interface Person {
  pet: Pet;
}
```

**Comment by bterlson** [comment, marked as answer]
I believe this should be coming soon as part of our Rest client emitter work, though I can't say for sure. We definitely intend to have "model emitters" in all languages we support eventually.

If it's interesting or useful, we have a test emitter that does this, which you could perhaps use. Just note that it's just for testing purposes and you will have to copy/paste the code. You can find it here: https://github.com/microsoft/typespec/blob/main/packages/compiler/test/emitter-framework/typescript-emitter.ts

**Reply by arabold** [reply]
Thank you for the reference. Indeed the example does what I was looking for 👍 Based on that I was able to create my own emitter that can create NestJS DTOs including class-validator decorators. Amazingly easy to use. I'm sold!

**Reply by bterlson** [reply]
Hey cool, welcome to the world of custom codegen 😁

Just to warn you, so you're not surprised - we are working on a revamp of the emitter framework to, among other things, better support taking a typescript interface emitter library like you copy/pasted and layer in special behavior (like creating NestJS DTOs), and that _may_ entail some amount of churn in these APIs. We will of course try not to break anything if at all possible.

**Reply by seidtgeist** [reply]
@bterlson That sounds super useful! Is it possible to follow the progress of that effort somewhere?

**Comment by crowbait** [comment]
Since this is the only reasonable Google result when searching for TS interfaces from TSP, I'll stick this here.

I have started work on a very simple emitter, which is capable of transforming at least basic TSP models to TypeScript interfaces.
It is currently quite tailored to what I need from it, but I'd be happy to work on it more and also it may be of some use for simpler projects out there, so I figured I'd share.

https://github.com/crowbait/typespec-typescript-emitter

It can also generate a TS object based on ops when using the HTTP library, which can help writing things like abstracted API wrappers (which is my specific use case and why I started on this in the first place). May be a bit specific, but still, might help someone.

**Reply by mcchrish** [reply]
There is also this https://www.npmjs.com/package/@azure-tools/typespec-ts. I don't use it personally but just found it while searching for typescript emitters.

**Reply by otomad** [reply]
> There is also this https://www.npmjs.com/package/@azure-tools/typespec-ts. I don't use it personally but just found it while searching for typescript emitters.

I tried this, however it generates extra REST API code.
